Ical.Net.Tests/MatchTimeZoneTests.cs

```diff
@@ -0,0 +1,242 @@
+            PRODID:-//Example Corp//NONSGML Event//EN
+            BEGIN:VEVENT
+            UID:example1
+            SUMMARY:Event with local time and time zone
+            DTSTART;TZID=America/New_York:20231101T090000
+            RRULE:FREQ=DAILY;UNTIL=20231105T130000Z
+            DTEND;TZID=America/New_York:20231101T100000
+            END:VEVENT
+            END:VCALENDAR
+            """;
+
+        var calendar = Calendar.Load(ical);
+        var evt = calendar.Events.First();
+        var until = evt.RecurrenceRules.First().Until;
+
+        var expectedUntil = new DateTime(2023, 11, 05, 13, 00, 00, DateTimeKind.Utc);
+        var occurrences = evt.GetOccurrences(new CalDateTime(2023, 11, 01), new CalDateTime(2023, 11, 06));
+
+        Assert.Multiple(() =>
+        {
+            Assert.That(until, Is.EqualTo(expectedUntil));
+            Assert.That(occurrences.Count, Is.EqualTo(4));
+            /*
+               Should have 4 occurrences:
+               November 1, 2023: 09:00 AM - 10:00 AM (UTC-0400) (America/New_York)
+               November 2, 2023: 09:00 AM - 10:00 AM (UTC-0400) (America/New_York)
+               November 3, 2023: 09:00 AM - 10:00 AM (UTC-0400) (America/New_York)
+               November 4, 2023: 09:00 AM - 10:00 AM (UTC-0400) (America/New_York)
+
+               November 5, 2023: 09:00 AM - 10:00 AM (UTC-0500) (America/New_York)
+               must NOT be included, because 20231105T130000Z => November 5, 2023: 08:00 AM (America/New_York)
+               (Daylight Saving Time in America/New_York ended on Sunday, November 5, 2023, at 2:00 AM)
+           */
+        });
+    }
+
+    [Test, Category("Recurrence")]
+    [TestCase("20241005T140000Z", 5)]
+    [TestCase("20241005T150000Z", 6)]
+    public void MatchTimeZone_LocalTimeAustraliaWithTimeZone(string inputUntil, int expectedOccurrences)
+    {
+        // DTSTART with local time and time zone reference (positive offset), UNTIL as UTC
+        var ical =
+           $"""
+            BEGIN:VCALENDAR
+            VERSION:2.0
+            PRODID:-//Example Corp//NONSGML Event//EN
+            BEGIN:VEVENT
+            UID:example1
+            SUMMARY:Event with local time and time zone
+            DTSTART;TZID=Australia/Sydney:20241001T010000
+            RRULE:FREQ=DAILY;UNTIL={inputUntil}
+            DTEND;TZID=Australia/Sydney:20241001T020000
+            END:VEVENT
+            END:VCALENDAR
+            """;
+
+        var calendar = Calendar.Load(ical);
+        var evt = calendar.Events.First();
+        var until = evt.RecurrenceRules.First().Until;
+
+        var expectedUntil = DateTime.ParseExact(inputUntil, "yyyyMMddTHHmmssZ",
+            System.Globalization.CultureInfo.InvariantCulture,
+            System.Globalization.DateTimeStyles.AssumeUniversal |
+            System.Globalization.DateTimeStyles.AdjustToUniversal);
+        var occurrences = evt.GetOccurrences(new CalDateTime(2024, 10, 01), new CalDateTime(2024, 10, 07));
```

Comment:
In this case DTSTART has a tzid but the startTime is floating. This is quite a special case and we should considere to disallow the case. Anyhow, I think it would be preferable to specify a tzid for the start/endTime too.

Reply:
Hm, CalDateTime(2024, 10, 01) resolves to a DATE value, which cannot have a timezone (RFC 5545)?

Reply:
So you mean, it should always be a DATE-TIME?

Reply:
Just a comment, not too relevant. Generally it would be advisable for users of the app, not to mix DATE vs DATE-TIME and floating vs non-floating, because there's always some ambiguity that causes additional complexity. In this case it really doesn't matter, so no need to change it. But as its advisable for users, not to mix it, we generally might want to avoid mixing them in the tests either (except for those cases where we explicitly want to test mixing them).